### PR TITLE
Event Type support 

### PR DESCRIPTION
**Pull request by matzew**
Signed-off-by: Matthias Wessendorf <mwessend@redhat.com>

Fixes #

## Proposed Changes

- adding `sourcecrd`'s _Controller_ to support "Even Type" (and required cluster roles)
-
-

**Release Note**

```release-note

```

**Docs**

### CODE REVIEW

File: cmd/source/controller/main.go [hunk 1]
```diff
@@ -196,5 +198,8 @@ func main() {
 		binding.NewController, NewKafkaBindingWebhook(kfkSelector),
 
 		source.NewController,
+
+		// Sources CRD
+		sourcecrd.NewController,
```
Comment (matzew) [inline cmd/source/controller/main.go:203]: @lionelvillard I fear/guess something like this must be done for all sources?

**Comment by matzew** [comment]
/hold
let's hold for reviews..

File: vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	cmw configmap.Watcher,
+) *controller.Impl {
+	crdInformer := crdinfomer.Get(ctx)
+
+	r := &Reconciler{
+		ogctx:       ctx,
+		ogcmw:       cmw,
+		controllers: make(map[schema.GroupVersionResource]runningController),
+	}
+	filterFunc := pkgreconciler.LabelFilterFunc(sources.SourceDuckLabelKey, sources.SourceDuckLabelValue, false)
```
Comment (pierDipi) [inline vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go:50]: Looking at this filter, it should work by changing the label key on the KafkaSource CRD.

https://github.com/knative-sandbox/eventing-kafka/blob/7a13f9656bd7aef3fe07ccfbf58a7fb509726b05/config/source/common/resources/source/kafkasource.yaml#L20

that should be `sources.knative.dev/source: "true"`

Comment (pierDipi) [reply vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go:50]: Label key: https://github.com/knative-sandbox/eventing-kafka/blob/2b185a9b59347cc4a16471ed0fb2ba3bdf6e1672/vendor/knative.dev/eventing/pkg/apis/sources/register.go#L30

Comment (matzew) [reply vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go:50]: look here:
https://github.com/knative/eventing/blob/main/config/core/resources/pingsource.yaml#L20

they all use `eventing.knative.dev/source` and the spec just mentions the `duck`:
https://github.com/knative/eventing/blob/main/docs/spec/sources.md#source-crds

Comment (matzew) [reply vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go:50]: so even I change that... I do not see any event type.... 

broken API ... :see_no_evil:

File: cmd/source/controller/main.go (hunk 1 above)
Comment (lionelvillard) [reply cmd/source/controller/main.go:203]: this shouldn't be need. This controller is already included in the eventing controller

Comment (lionelvillard) [reply cmd/source/controller/main.go:203]: I'm getting this in the eventing controller log, when setting the proper duck type label on the KafkaSource CRD:

```
{"level":"error","ts":"2021-12-07T15:38:50.632Z","logger":"controller","caller":"duck/duck.go:78","msg":"runtime object is not convertible to Source duck type: ","commit":"d5201e7","knative.dev/pod":"eventing-controller-c6c8b75dd-zh9xs","knative.dev/traceid":"fa0b199c-9bff-42bf-a99d-77683789b0a6","knative.dev/key":"sources-kafka-sanity/fruits","runtimeObj":null,"stacktrace":"knative.dev/eventing/pkg/reconciler/source/duck.(*Reconciler).Reconcile\n\tknative.dev/eventing/pkg/reconciler/source/duck/duck.go:78\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20211111114938-0b0c3390a475/controller/controller.go:542\nknative.dev/pkg/controller.(*Impl).RunContext.func3\n\tknative.dev/pkg@v0.0.0-20211111114938-0b0c3390a475/controller/controller.go:478"}
{"level":"info","ts":"2021-12-07T15:38:50.632Z","logger":"controller","caller":"controller/controller.go:550","msg":"Reconcile succeeded","commit":"d5201e7","knative.dev/pod":"eventing-controller-c6c8b75dd-zh9xs","knative.dev/traceid":"fa0b199c-9bff-42bf-a99d-77683789b0a6","knative.dev/key":"sources-kafka-sanity/fruits","duration":0.000156068}
```

Comment (lionelvillard) [reply cmd/source/controller/main.go:203]: I'm guessing this is a RBAC issue.

File: vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go (hunk 2 above)
Comment (lionelvillard) [reply vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go:50]: this is messy. The aggregates ClusterRole expects `duck.knative.dev/source: "true"`

Comment (matzew) [reply vendor/knative.dev/eventing/pkg/reconciler/source/crd/controller.go:50]: we have `duck.knative.dev/source: "true"` already - today:
https://github.com/knative-sandbox/eventing-kafka/blob/main/config/source/common/resources/source/kafkasource.yaml#L21

File: cmd/source/controller/main.go (hunk 1 above)
Comment (matzew) [reply cmd/source/controller/main.go:203]: /hold
let's not merge... but not sure on actual detail, atm

**Pull request closed without merging**